### PR TITLE
Update Eclipselink to 4.0.1 for JPA 3.1 feature

### LINKED ISSUE #23952: [JPA 3.1] EclipseLink: Deliver Issue #1779
Update JPA 3.1 to fix EclipseLink issue [1779](https://github.com/eclipse-ee4j/eclipselink/issues/1779)

### PR DESCRIPTION
**Pull request by jgrassel**
fixes
#23952
#23944
#23680
#23566

**Note: Do not merge until after 23.0.0.3 GAs**, since this update has ID changes that will need translation.

Updating the Eclipselink associated with the `persistence-3.1` feature to 4.0.1.  This includes the noteworthy "ASM Service", which allows for choosing between either the bundled Eclipselink ASM implementation, or the regular OW2 ASM implementation.

The default ASM implementation that is used is still the Eclipselink ASM.  To instruct Eclipselink to use the OW2 ASM, a user needs to add `-Declipselink.asm.service=ow2` to the server's `jvm.options` file.

FAT tests have been added to test this new capability added in Eclipselink 4.0.1.

Message updates back to Eclipselink for 4.0.2:
https://github.com/eclipse-ee4j/eclipselink/pull/1820
https://github.com/eclipse-ee4j/eclipselink/pull/1861

The validateNLS excludes added by this PR should be removed in 4.0.2 (confirm 1861 has been merged first)

for #27501

### CODE REVIEW

**Comment by jgrassel** [comment]
#run-libby-bot

**Comment by wraschke** [comment]
#pb

**Comment by wraschke** [comment]
#pbbeta

**Comment by wraschke** [comment]
#build

**Comment by posmith** [comment]
L2 message review completed

**Comment by jgrassel** [comment]
#build

File: dev/io.openliberty.persistence.3.1.thirdparty/resources/org/eclipse/persistence/exceptions/i18n/ValidationExceptionResource.nlsprops [hunk 1]
```diff
@@ -633,5 +633,9 @@
 
 7356=Procedure: [{1}] cannot be executed because {0} does not currently support multiple out parameters
 
-7357=The "[{0}]" URL for "[{1}]" resource does not belong to a valid persistence root (as per Jakarta Persistence Specification)
+7357=URL [{0}] for resource [{1}] does not belong to a valid persistence root (as per JPA Specification)
```
Comment (dmuelle) [inline dev/io.openliberty.persistence.3.1.thirdparty/resources/org/eclipse/persistence/exceptions/i18n/ValidationExceptionResource.nlsprops]:
```suggestion
7357=The [{0}] URL for the [{1}] resource does not belong to a valid persistence root, as defined by the JPA Specification.
```

File: dev/io.openliberty.persistence.3.1.thirdparty/resources/org/eclipse/persistence/internal/localization/i18n/LoggingLocalizationResource.nlsprops [hunk 2]
```diff
@@ -115,6 +115,10 @@ connected_user_database=Connected: {3}{4}User: {0}{3}{4}Database: {1}  Version:
 
 connected_user_database_driver=Connected: {0}{6}User: {1}{6}Database: {2}  Version: {3}{6}Driver: {4}  Version: {5}
 
+corrupt_object=corrupt object: {0}
+
+corrupt_object_referenced_through_mapping=corrupt object referenced through mapping: {0}
```
Comment (dmuelle) [inline dev/io.openliberty.persistence.3.1.thirdparty/resources/org/eclipse/persistence/internal/localization/i18n/LoggingLocalizationResource.nlsprops]:
```suggestion
corrupt_object_referenced_through_mapping=The following corrupt object is referenced through mapping: {0}.
```

File: dev/io.openliberty.persistence.3.1.thirdparty/resources/org/eclipse/persistence/internal/localization/i18n/LoggingLocalizationResource.nlsprops [hunk 3]
```diff
@@ -717,6 +723,8 @@ validate_db_schema_with_not_supported=WLS native CMP setting "validate-db-schema
 
 validate_ejb_jar=Validating ejb-jar.xml starts, it may take some times...
 
+validate_object_space=validate object space.
+
 verifiy_columns_read_locking_not_supported=Optimistic setting "Read" on "verify-columns" in entity ({0}) is not directly supported in EclipseLink CMP. See migration doc for details.
 
 verifiy_rows_read_locking_not_supported=Optimistic setting "Read" on "verify-rows" in entity ({0}) is not directly supported in EclipseLink CMP. See migration doc for details..
```
Comment (dmuelle) [inline dev/io.openliberty.persistence.3.1.thirdparty/resources/org/eclipse/persistence/internal/localization/i18n/LoggingLocalizationResource.nlsprops]:
```suggestion
verifiy_rows_read_locking_not_supported=Optimistic setting "Read" on "verify-rows" in entity ({0}) is not directly supported in EclipseLink CMP. See migration doc for details.
```

**Review by dmuelle — CHANGES_REQUESTED** [review]
ID review, let me know if you have any questions, thanks

File: dev/io.openliberty.persistence.3.1.thirdparty/resources/org/eclipse/persistence/internal/localization/i18n/LoggingLocalizationResource.nlsprops [hunk 4]
```diff
@@ -717,9 +723,11 @@ validate_db_schema_with_not_supported=WLS native CMP setting "validate-db-schema
 
 validate_ejb_jar=Validating ejb-jar.xml starts, it may take some times...
 
+validate_object_space=validate object space.
+
 verifiy_columns_read_locking_not_supported=Optimistic setting "Read" on "verify-columns" in entity ({0}) is not directly supported in EclipseLink CMP. See migration doc for details.
 
-verifiy_rows_read_locking_not_supported=Optimistic setting "Read" on "verify-rows" in entity ({0}) is not directly supported in EclipseLink CMP. See migration doc for details..
+verifiy_rows_read_locking_not_supported=Optimistic setting "Read" on "verify-rows" in entity ({0}) is not directly supported in EclipseLink CMP. See migration doc for details.
```
Comment (dmuelle) [inline dev/io.openliberty.persistence.3.1.thirdparty/resources/org/eclipse/persistence/internal/localization/i18n/LoggingLocalizationResource.nlsprops:730]:
```suggestion
verifiy_rows_read_locking_not_supported=Optimistic setting "Read" on "verify-rows" in the  ({0}) entity is not directly supported in EclipseLink CMP. See migration documentation for details.
```

Comment (dmuelle) [reply dev/io.openliberty.persistence.3.1.thirdparty/resources/org/eclipse/persistence/internal/localization/i18n/LoggingLocalizationResource.nlsprops:730]: Missed this the first time, other changes LGTM

**Review by dmuelle — DISMISSED** [review]
LGTM thanks

**Comment by jgrassel** [comment]
#build

**Comment by jgrassel** [comment]
#build

**Comment by jgrassel** [comment]
Looking at the build analysis (which hasn't been deleted), looks like just 2 buckets failed and they weren't related to JPA.  Think this safe to merge once a review is done.

**Review by dazey3 — APPROVED** [review]
LGTM

**Comment by jgrassel** [comment]
#run-libby-bot

**Pull request merged**